engines/illusions/bbdou/bbdou_credits.cpp

```diff
@@ -0,0 +1,246 @@
+/* ScummVM - Graphic Adventure Engine
+ *
+ * ScummVM is the legal property of its developers, whose names
+ * are too numerous to list here. Please refer to the COPYRIGHT
+ * file distributed with this source distribution.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ */
+
+#include "illusions/bbdou/illusions_bbdou.h"
+#include "illusions/bbdou/bbdou_credits.h"
+#include "illusions/actor.h"
+#include "illusions/dictionary.h"
+#include "illusions/textdrawer.h"
+#include "illusions/time.h"
+#include "illusions/resources/scriptresource.h"
+
+namespace Illusions {
+
+BbdouCredits::BbdouCredits(IllusionsEngine_BBDOU *vm)
+	: _vm(vm) {
+}
+
+BbdouCredits::~BbdouCredits() {
+}
+
+void BbdouCredits::start(uint32 endSignalPropertyId, float speedModifier) {
+	_endSignalPropertyId = endSignalPropertyId;
+	_currFontId = 0x120004;
+	_currLineIndex = 1;
+	_split = false;
+	// convertTextData();
+	initCreditsItems();
+	createCreditsThread(speedModifier);
+}
+
+void BbdouCredits::stop() {
+	freeCreditsItems();
+}
+
+void BbdouCredits::drawNextLine() {
+	uint leftIndex, rightIndex;
+	
+	if (!readNextLine(leftIndex, rightIndex)) {
+		_vm->_scriptResource->_properties.set(_endSignalPropertyId, true);
+		return;
+	}
+
+	if (leftIndex) {
+		const char *leftText = getText(leftIndex);
+		if (leftText && strlen(leftText) != 0) {
+			uint32 objectId = getNextFreeObjectId();
+			int alignment = rightIndex ? 1 : 2;
+			drawTextToControl(objectId, leftText, alignment);
+		}
+	}
+
+	if (rightIndex) {
+		const char *rightText = getText(rightIndex);
+		if (rightText && strlen(rightText) != 0) {
+			uint32 objectId = getNextFreeObjectId();
+			drawTextToControl(objectId, rightText, 4);
+		}
+	}
+
+}
+
+void charToWChar(const char *text, uint16 *wtext, uint size) {
```

Comment:
I don't think the text is in utf8 format here. I think the widening from 8 to 16bits is to fit in with the other dialog strings that are 16 bits.

Reply:
It is not UTF8, but UTF16. This is then fed to the TTF renderer.

Reply:
It's simply converts byte to uint16 per char since the text functions all use 16bits per character (don't ask me why).
It's not Unicode/UTF-8. Duckman doesn't use TTF.